CL-3548 - Scaffold public api v2

jamesspeake opened the pull request:
# Changelog
- New endpoints for Havant

jamesspeake committed:
[CL-3548] Set up generation of openapi specs with a swagger UI container to test them

jamesspeake committed:
[CL-3548] Updated existing controllers and routes for project, phase and ideas

jamesspeake committed:
[CL-3548] Removed debug

jamesspeake committed:
[CL-3548] Added user endpoint & further descriptions for open api

jamesspeake committed:
[CL-3548] Created more shared code, /:id endpoints and added proper open_api description

jamesspeake committed:
[CL-3548] Rubocop offenses

jamesspeake committed:
[CL-3548] Minor simplification

jamesspeake committed:
[CL-3548] Refactored so all new endpoints are in /v2 and locale is a normal query parameter

jamesspeake committed:
[CL-3548] Refactored so all new endpoints are in /v2 and locale is a normal query parameter

jamesspeake committed:
[CL-3548] Removed comment

jamesspeake committed:
[CL-3548] Added all phases endpoint

jamesspeake committed:
[CL-3548] Added all phases endpoint

jamesspeake committed:
[CL-3548] Added standard filters and tests to all controllers

jamesspeake commented on back/engines/commercial/multi_tenancy/spec/factories/tenants.rb:
```diff
@@ -36,13 +36,13 @@
           organization_name: {
             'en' => 'Liege',
-            'nl-NL' => 'Luik',
+            'nl-BE' => 'Luik',
```
Changed this as most of the factories produce multilocs with nl-BE not nl-NL. Having to fix a few tests as a result.

jamesspeake committed:
[CL-3548] Rolled back

jamesspeake committed:
[CL-3548] Added both nl-NL and nl-BE to the test tenant config

jamesspeake committed:
Merge branch 'master' into CL-3548-scaffold-public-api-v2

jamesspeake committed:
[CL-3548] Rubocop

jamesspeake committed:
[CL-3548] Standardise tenant locales

jamesspeake committed:
[CL-3548] Added comments

jamesspeake committed:
[CL-3548] Temp fix for locales in tests

jamesspeake committed:
Merge branch 'master' into CL-3548-scaffold-public-api-v2

jamesspeake committed:
[CL-3548] Added votes controller, TODOs and factored out common functions into the main controller

jamesspeake committed:
[CL-3548] Added votes controller, TODOs and factored out common functions into the main controller

jamesspeake committed:
[CL-3548] Added initiatives

jamesspeake committed:
[CL-3598] Added initiative comments/votes

jamesspeake committed:
[CL-3598] Rubocop fixes

jamesspeake committed:
[CL-3598] Changed a few TODOs

adessy committed:
[CL-3548] Refactor routes.rb

adessy committed:
[CL-3548] Remove instance class var from comment_serializer.rb

adessy committed:
[CL-3548] Remove instance class var from idea_serializer.rb

adessy committed:
[CL-3548] Remove instance class var from initiative_serializer.rb

adessy committed:
[CL-3548] Remove instance class var from phase_serializer.rb

adessy committed:
[CL-3548] Remove instance class var from project_serializer.rb

adessy committed:
[CL-3548] Remove instance class var from topic_serializer.rb

adessy committed:
[CL-3548] Remove instance class var from user_serializer.rb

adessy committed:
[CL-3548] Refactor ideas_controller.rb

adessy committed:
[CL-3548] Refactor initiatives_controller.rb

adessy committed:
[CL-3548] Refactor phases_controller.rb

adessy committed:
[CL-3548] Refactor projects_controller.rb

adessy committed:
[CL-3548] Refactor users_controller.rb

adessy committed:
[CL-3548] Use the same serializer for all comments

Use the same serializer for all comment resources, regardless of the route from which they are returned. I think this will make things easier for both us and the client by not having to deal with different cases.

adessy committed:
[CL-3548] Use the same serializer for all votes

Use the same serializer for all vote resources, regardless of the route from which they are returned. I think this will make things easier for both us and the client by not having to deal with different cases.

adessy committed:
[CL-3548] Remove newly added exceptions from .rubocop_todo.yml

adessy committed:
[CL-3548] Clean comment

adessy committed:
[CL-3548] Rework routes.rb

This commit reworks the route definitions using the `collection` method to ensure that in `v2/ideas/comments` (and similar routes), `comments` is not mistakenly considered as an `:id`.

adessy committed:
Merge branch 'master' into CL-3548-scaffold-public-api-v2

adessy reviewed (commented):
Hey @jamesspeake , what do you think about replacing the following endpoints:
- `api/v2/ideas/comments/votes`
- `api/v2/initiatives/comments/votes`
- `api/v2/comments/votes`
- `api/v2/iniatives/votes`
- `api/v2/ideas/votes`

with something like `api/v2/votes?votable={votable}`, where `votable` can be any of the following: `idea`, `initiative`, `comment`, `initiative-comment`, `idea-comment`?

I think it works pretty well for `idea`, `initiative`, and `comment`; a bit less well for `initiative-comment` and `idea-comment` since those are not actual values of the `votable` attribute. However, overall, I believe this approach has the benefits of:
- helping to keep the number of endpoints under control,
- being somewhat more standard, perhaps.

(The `votable` query parameter could be given another name as it might not be the most user-friendly choice.)

Finally, this is just an example, and this line of thought could be applied more broadly to the entire API specification.

adessy committed:
[CL-3548] Refactor the various /votes routes into a single route

Replace the following routes:
- /public_api/v2/ideas/votes
- /public_api/v2/ideas/comments/votes
- /public_api/v2/initiatives/votes
- /public_api/v2/initiatives/comments/votes
- /public_api/v2/comments/votes

with a single route:
/public_api/votes
that has a query parameter "votable_type" to retrieve votes for specific types of resources.

adessy committed:
[CL-3548] Implement BaseSerializer for public API v2

This commit adds a BaseSerializer that includes helper/utility methods for converting between classname-like strings and the dasherized notation used by the API.

jamesspeake committed:
Merge branch 'master' into CL-3548-scaffold-public-api-v2

# Conflicts:
#	back/engines/commercial/public_api/app/controllers/public_api/public_api_controller.rb
#	back/spec/factories/votes.rb

jamesspeake committed:
[CL-3548] Refactored votes to reactions

adessy committed:
[CL-3548] Refactor the /comments routes into a single route

Replace the following routes:
- /public_api/v2/ideas/comments
- /public_api/v2/initiatives/comments

with a single route:
/public_api/comments
that has a query parameter "post_type" to retrieve comments on specific types of resources.

adessy committed:
Merge remote-tracking branch 'origin/CL-3548-scaffold-public-api-v2' into CL-3548-scaffold-public-api-v2

# Conflicts:
#	back/engines/commercial/public_api/app/controllers/public_api/v2/votes_controller.rb
#	back/engines/commercial/public_api/app/serializers/public_api/v2/vote_serializer.rb

adessy committed:
[CL-3548] Rework ideas_spec.rb

adessy committed:
[CL-3548] Rework initiatives_spec.rb

adessy committed:
[CL-3548] Use around_action to switch locale for public API

From doc:
> I18n.locale can leak into subsequent requests served by the same thread/process if it is not consistently set in every controller. For example executing I18n.locale = :es in one POST requests will have effects for all later requests to controllers that don't set the locale, but only in that particular thread/process. For that reason, instead of I18n.locale = you can use I18n.with_locale which does not have this leak issue.

adessy committed:
[CL-3548] Rework phases_spec.rb

adessy committed:
[CL-3548] Rework users_spec.rb

adessy committed:
[CL-3548] Fix: Remove before_action :set_locale hook

`set_locale` was replaced by `switch_locale` in 48427f88d6a4b67a51e0188edcfb342ddcca4210

adessy committed:
[CL-3548] Rework topics_spec.rb

adessy committed:
[CL-3548] Fix phase test

adessy committed:
[CL-3548] Rework projects_spec.rb

adessy committed:
[CL-3548] Fix topics_spec.rb

adessy committed:
[CL-3548] Fix rubocop offenses

adessy commented:
Hey @jamesspeake! Here we are, ready for a review! In terms of scope, I didn't add anything with respect to the original PR, but I reworked a few parts, most notably:
- merging some serializers (vote serializers and comment serializers),
- merging some routes as we discussed (see https://github.com/CitizenLabDotCo/citizenlab/pull/4851#pullrequestreview-1474532235),
- reworking the OpenAPI doc a bit,
- fixing issues with localization (see https://github.com/CitizenLabDotCo/citizenlab/pull/4851/commits/48427f88d6a4b67a51e0188edcfb342ddcca4210).

Once we're done with this one, I'll move on to expand these newly-introduced endpoints (adding parameters, documenting the response model, etc.) and adding the missing ones. But I would also like to question the necessity of the 'locale' parameter. It seems to me that this parameter actually complicates matters for both end users and ourselves. For instance:
- It makes it difficult to identify resources (such as ideas) that have multiple translations.
- Users are required to send multiple requests if they want to retrieve all translations for a specific resource.
- It obscures the locale of the response, as a fallback translation is used if the specified locale is not available, and users have no way to tell.

I would suggest dropping it and returning the multilocs instead.

jamesspeake commented on back/engines/commercial/public_api/app/controllers/public_api/public_api_controller.rb:
```diff
@@ -3,20 +3,84 @@
+
+    # Default per page is 12, maximum is 24
+    def num_per_page
+      [params[:page_size]&.to_i || 12, 24].min
```
Let's double the max to 48. Don't think it will harm.

adessy replied:
Yeah, that's something I intended to change. Is there any specific reason why you like multiples of 12? :)

jamesspeake replied:
No reason - given that the current values were 12 & 24 - I thought maybe we had some kind of convention to use nice binary numbers!

jamesspeake commented:
> It seems to me that this parameter actually complicates matters for both end users and ourselves

I think we should keep it as is. For use within other platforms it keeps it easier to manipulate - given that the current use case is for a single language platform - having to extract text from a multiloc is an extra step that shouldn't be required. We could always add the multiloc fields in later given that the fields we are returning are minus the _multiloc

jamesspeake commented on back/engines/commercial/public_api/app/controllers/public_api/public_api_controller.rb:
```diff
@@ -17,17 +51,18 @@ def meta_properties(relation)
       dates = date_values.split(',')
-      start_date = (dates[0].presence || '2017-01-01')
-      end_date = (dates[1].presence || Date.tomorrow.strftime('%F'))
+      # TODO: fix SQL injection vulnerability?
```
Do we need to do anything with this right now - fairly low risk for now I guess if not

adessy replied:
I'd rather keep it for now. I'd like to think a bit more about the severity of the issue

jamesspeake commented on back/engines/commercial/public_api/spec/acceptance/v2/users_spec.rb:
```diff
@@ -5,78 +5,87 @@
       let(:page_size) { 2 }
 
-      example_request 'Successful response' do
+      example_request 'List only the first page of users' do
```
Have to check whether 'test criteria' read correctly when the open api documentation is produced - I'd focused my tests on what the api docs looked like rather than describing the tests well! Not an issue though, it can always be changed if it doesn't read so well.

jamesspeake reviewed (commented):
All looks good to me apart from a couple of comments.

adessy committed:
[CL-3548] Increase page sizes
